src/http/server.cr

```diff
@@ -95,90 +100,138 @@ class HTTP::Server
+  # Creates a `TCPServer` listenting on `127.0.0.1:port` and adds it as a socket,
+  # returning the local address and port the server listens on.
+  #
+  # If *reuse_port* is `true`, it enables the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind_tcp(port : Int32, reuse_port : Bool = false) : Socket::IPAddress
+    bind_tcp "127.0.0.1", port, reuse_port
   end
 
-  # Returns the TCP port the server is connected to.
+  # Creates a `TCPServer` listening on an unused port and adds it as a socket.
+  #
+  # Returns the `Socket::IPAddress` with the determined port number.
   #
-  # For example you may let the system choose a port, then report it:
   # ```
-  # server = HTTP::Server.new(0) { }
-  # server.bind
-  # server.port # => 12345
+  # server = HTTP::Server.new { }
+  # server.bind_unused_port # => Socket::IPAddress.new("127.0.0.1", 12345)
   # ```
-  def port
-    if server = @server
-      server.local_address.port.to_i
-    else
-      @port
+  def bind_unused_port(host : String = "127.0.0.1", reuse_port : Bool = false) : Socket::IPAddress
+    bind_tcp host, 0, reuse_port
+  end
+
+  # Adds a `Socket::Server` *socket* to this server.
+  def bind(socket : Socket::Server) : Nil
+    @sockets << socket
+  end
+
+  # Enumerates all addresses this server is bound to.
+  def each_address(&block : Socket::Address ->)
+    @sockets.each do |socket|
+      yield socket.local_address
+    end
+  end
+
+  def addresses : Array(Socket::Address)
+    array = [] of Socket::Address
+    each_address do |address|
+      array << address
     end
+    array
   end
 
-  # Creates the underlying `TCPServer` if the doesn't already exist.
+  # Creates a `TCPServer` listenting on `127.0.0.1:port`, adds it as a socket
+  # and starts the server. Blocks until the server is closed.
   #
-  # You may set *reuse_port* to true to enable the `SO_REUSEPORT` socket option,
-  # which allows multiple processes to bind to the same port.
-  def bind(reuse_port = false)
-    @server ||= TCPServer.new(@host, @port, reuse_port: reuse_port)
+  # See `#bind(port : Int32)` for details.
+  def listen(port : Int32, reuse_port : Bool = false)
+    bind_tcp(port, reuse_port)
+
+    listen
   end
 
-  # Starts the server. Blocks until the server is closed.
+  # Creates a `TCPServer` listenting on `host:port`, adds it as a socket
+  # and starts the server. Blocks until the server is closed.
   #
-  # See `#bind` for details on the *reuse_port* argument.
-  def listen(reuse_port = false)
-    server = bind(reuse_port)
-    until @wants_close
-      spawn handle_client(server.accept?)
+  # See `#bind(host : String, port : Int32)` for details.
+  def listen(host : String, port : Int32, reuse_port : Bool = false)
+    bind_tcp(host, port, reuse_port)
+
+    listen
+  end
+
+  # Starts the server. Blocks until the server is closed.
+  def listen
+    raise "Can't start server with not sockets to listen to" if @sockets.empty?
```

Comment:
Maybe we should include something like "Use HTTP::Server#bind" in the error so we can guide the user a bit. WDYT?

Reply:
Another issue, which should probably also raise, is when `bind` is called after `listen`:
```cr
server.bind socket
spawn { server.listen }
server.bind socket2
```

For this we'd need a ivar to track if the server is listening. It could probably be implemented to retroactively start listening on the additional socket as well, but that's more complicated and probably not that useful without having service objects.

Reply:
Yeap, I would say to add an ivar to track that too.